### ISSUE TITLE
Deep-cover seems to break a test.

### ISSUE THREAD

**Issue by contribu** [deep-cover/deep-cover#47]
Thank you for creating a great gem.
But, I found a bug that seems to be caused by the ruby bug.
The bug can be reproduced in ruby 2.5.1 and cannot be reproduced in ruby 2.6.0.

I have confirmed that it works with ruby 2.6.0, so I am not too much trouble.
However, I report because I investigated it.

I think fixing this bug in ruby 2.5.1 is difficult so it may be good to add a test to know under which ruby version deep-cover works or to write caveats.
(Sorry for leaving all the work but I have no time to complete it)

## Environment

```
ruby --version
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]
```

## Bug overview

The following code is broken when deep-cover is enabled.
It seems to occur when the number of arguments is too large.

```ruby
class NgClass
  def self.my_class_method(
      a0: 0, a1: 0, a2: 0, a3: 0, a4: 0, a5: 0, a6: 0, a7: 0, a8: 0, a9: 0,
      b0: 0, b1: 0, b2: 0, b3: 0, b4: 0, b5: 0, b6: 0, b7: 0, b8: 0, b9: 0,
      c0: 0, c1: 0, c2: 0, c3: 0, c4: 0, c5: 0, c6: 0, c7: 0, c8: 0, c9: 0,
      d0: 0, d1: 0,
      x: 0
    )
    x
  end
end
```

## Rewritten version of above (broken even without deep-cover)

```ruby
($_cov ||= {})[0]||=Array.new(38,0);$_cov[0][37]+=1;_temp=nil;((_temp=(class NgClassRewritten
  $_cov[0][36]+=1;((_temp=(def (self).my_class_method(
    a0: ($_cov[0][0]+=1;0), a1: ($_cov[0][1]+=1;0), a2: ($_cov[0][2]+=1;0), a3: ($_cov[0][3]+=1;0), a4: ($_cov[0][4]+=1;0), a5: ($_cov[0][5]+=1;0), a6: ($_cov[0][6]+=1;0), a7: ($_cov[0][7]+=1;0), a8: ($_cov[0][8]+=1;0), a9: ($_cov[0][9]+=1;0),
        b0: ($_cov[0][10]+=1;0), b1: ($_cov[0][11]+=1;0), b2: ($_cov[0][12]+=1;0), b3: ($_cov[0][13]+=1;0), b4: ($_cov[0][14]+=1;0), b5: ($_cov[0][15]+=1;0), b6: ($_cov[0][16]+=1;0), b7: ($_cov[0][17]+=1;0), b8: ($_cov[0][18]+=1;0), b9: ($_cov[0][19]+=1;0),
        c0: ($_cov[0][20]+=1;0), c1: ($_cov[0][21]+=1;0), c2: ($_cov[0][22]+=1;0), c3: ($_cov[0][23]+=1;0), c4: ($_cov[0][24]+=1;0), c5: ($_cov[0][25]+=1;0), c6: ($_cov[0][26]+=1;0), c7: ($_cov[0][27]+=1;0), c8: ($_cov[0][28]+=1;0), c9: ($_cov[0][29]+=1;0),
        d0: ($_cov[0][30]+=1;0), d1: ($_cov[0][31]+=1;0),
        x: ($_cov[0][32]+=1;0)
    )
    $_cov[0][34]+=1;_temp=nil;x
  end);$_cov[0][33]+=1;_temp=_temp))
end);$_cov[0][35]+=1;_temp=_temp))
```

## How to reproduce?

Please try this example.
https://github.com/contribu/deep_cover_test_fail

## Facts

- Ok when the number of arguments is 31
- Ng when the number of arguments is 32
- Ng when the class is rewritten by hand

## Hypothesis

Rewriting seems to be correct and the rewritten class doesn't pass the test without deep-cover, so, it may be a ruby bug.

## In Ruby 2.6.0

The above is not reproduced even when the number of arguments is 100.

**Comment by MaxLap** [comment]
Thank you for the report. I can confirm i can run your reproduction case and it fails on ruby 2.5 (and also on 2.4, didn't check before that), but not on 2.6. 

I will take a look tonight or tomorrow.

Good job on the investigation!

**Comment by MaxLap** [comment]
Interesting and quite so weird. Here are some notes I'm leaving myself and those interested. I will clean up my reproduction code.

To clarify a little bit, in 2.5.0 to 2.5.5, if a method takes more than 32 (so at least 33) keyword arguments that need their default value to be evaluated, then there is a problem where the default_value are often not evaluated and worse, often, they are only evaluated (and used) for the keyword arguments which received a value and so shouldn't be evaluated.

When I say "need to be evaluated", i mean that ruby integers and symbols do not count for this behavior, probably because they don't need to be evaluated since they are immutable so ruby has optimizations.

There is also another bug that is way more specific, that happens in 2.3, 2.4 and 2.5. 
When one of the passed argument is the last keyword argument that needs to be evaluated of the method and the method has between `33 and 32 + number of passed keyword arguments matching those that need to be evaluated` keyword arguments that need to be evaluated, then it seems the keyword arguments that need to be evaluated that are after the 32th will ignore the received argument and still use their default.

This all seems to be solved in 2.6.0. It's a very odd situation and the trivial fix for DeepCover we will probably do is to skip the handling of default args for keyword args when there are more than 32 for rubies < 2.6.

Thanks again

**Comment by MaxLap** [comment]
Just pushed a fix for this. It will be in the next release. (0.7.7)
